vernac/vernacentries.ml

```diff
@@ -2557,121 +2624,17 @@ let () =
-  try
-    (* If the command actually works, ignore its effects on the state.
-       * Note that error has to be printed in the right state, hence
-       * within the purified function *)
-    try let _ = f () in raise HasNotFailed
-    with
-    | HasNotFailed as e -> raise e
-    | e ->
-      let e = CErrors.push e in
-      raise (HasFailed (CErrors.iprint
-                          (ExplainErr.process_vernac_interp_error ~allow_uncaught:false e)))
-  with e when CErrors.noncritical e ->
-    (* Restore the previous state XXX Careful here with the cache! *)
-    Vernacstate.invalidate_cache ();
-    Vernacstate.unfreeze_interp_state st;
-    let (e, _) = CErrors.push e in
-    match e with
-    | HasNotFailed ->
-      user_err ~hdr:"Fail" (str "The command has not failed!")
-    | HasFailed msg ->
-      if not !Flags.quiet || !test_mode then Feedback.msg_info
-          (str "The command has indeed failed with message:" ++ fnl () ++ msg)
-    | _ -> assert false
-
-let interp ?(verbosely=true) ?proof ~st {CAst.loc;v=c} : Proof_global.t option =
-  let rec control ~st = function
-  | VernacExpr (atts, v) ->
-    aux ~atts ~st v
-  | VernacFail v ->
-    with_fail ~st (fun () -> ignore(control ~st v));
-    st.Vernacstate.proof
-  | VernacTimeout (n,v) ->
-    current_timeout := Some n;
-    control ~st v
-  | VernacRedirect (s, {v}) ->
-    Topfmt.with_output_to_file s (control ~st) v
-  | VernacTime (batch, ({v} as com)) ->
-    let header = if batch then Topfmt.pr_cmd_header com else Pp.mt () in
-    System.with_time ~batch ~header (control ~st) v;
-
-  and aux ~atts ~st : _ -> Proof_global.t option =
-    function
-
-    | VernacLoad (_,fname) ->
-      unsupported_attributes atts;
-      vernac_load ~st control fname
-
-    | c ->
-      (* NB: we keep polymorphism and program in the attributes, we're
-         just parsing them to do our option magic. *)
-      try
-        vernac_timeout begin fun st ->
-          let pstate : Proof_global.t option =
-            if verbosely
-            then Flags.verbosely (interp ?proof ~atts ~st) c
-            else Flags.silently  (interp ?proof ~atts ~st) c
-          in
-          pstate
-          end st
-      with
-      | reraise when
-          (match reraise with
-           | Timeout -> true
-           | e -> CErrors.noncritical e)
-        ->
-        let e = CErrors.push reraise in
-        let e = locate_if_not_already ?loc e in
-        let () = restore_timeout () in
-        iraise e
-  in
-  if verbosely
-  then Flags.verbosely (control ~st) c
-  else (control ~st) c
-
 (* Be careful with the cache here in case of an exception. *)
-let interp ?verbosely ?proof ~st cmd =
+let interp ?(verbosely=true) ?proof ~st cmd =
   Vernacstate.unfreeze_interp_state st;
-  try
-    let pstate = interp ?verbosely ?proof ~st cmd in
-    Vernacstate.Proof_global.set pstate;
-    Vernacstate.freeze_interp_state ~marshallable:false
+  try vernac_timeout ?timeout:!default_timeout (fun st ->
```

Comment:
If we have 
~~~coq
Set Default Timeout 1.
Timeout 3 Check thing_taking_2s.
~~~
it seems like it won't work properly.

Reply:
Indeed that's a change of semantics, well spotted. That being said, doesn't the new behavior make more sense?

Reply:
Like what should happen in `Timeout 1 (Timeout 2 foo))` ?

Reply:
`Error: multiple timeouts not allowed`
I'm concerned about the specific one I posted with Default Timeout, where it makes sense for a local Timemout to override it.

Reply:
Could make sense otherwise [the global taking priority over the local] but I'll try to restore behavior.

Reply:
>the global taking priority over the local

Not with it being called Default imo

Reply:
Ok I tried to amend the change.

Reply:
>Ok I tried to amend the change.

Did you test something like the example I posted? USe something like
~~~coq
Set Default Timeout 1.

Lemma foo : True.
  Timeout 2 do 2000 pose I.
~~~

Reply:
Yup it seems to work.

Reply:
Cool.